File: Golang/p39_Combination_sum.go/combination_sum.go
package main

import (
	"fmt"
	"sort"
)

// Input (target = 7 and candicates or nums  [2,3,6,7])
// The element in output can be duplicate as long as the total of List == target
// Output :[[2,2,3],[7]]

func combinationSum(candicates []int, target int) [][]int {
	res, sub := [][]int{}, []int{}
	//Sort array
	sort.Ints(candicates)

	var backtracking func(i int, sub []int, target int, candicates []int, res *[][]int, total int)
	backtracking = func(i int, sub []int, target int, candicates []int, res *[][]int, total int) {
		// if total == target then return output
		if total == target {
			temp := make([]int, len(sub))
			copy(temp, sub)
			*res = append(*res, temp)
			return
		}
		//Return back to previous case
		if i >= len(candicates) || total > target {
			return
		}
		sub = append(sub, candicates[i])
		backtracking(i, sub, target, candicates, res, total+candicates[i])
		//Pop
		sub = sub[:len(sub)-1]
		//Skip the current candidate and move to the n
		backtracking(i+1, sub, target, candicates, res, total)
	}
	backtracking(0, sub, target, candicates, &res, 0)
	return res
}
func main() {
	fmt.Println(combinationSum([]int{2, 3, 6, 7}, 7))
}

File: Python/P39_Combination_sum/Combination_sum.py
#Input (target = 7 and candicates or nums  [2,3,6,7])
#The element in output can be duplicate as long as the total of List == target
#Output :[[2,2,3],[7]]

from typing import List
class Solution:
    def combinationSum(self, candicates: List[int], target: int) -> List[List[int]]:
        result = []
        #Total is sum of ele in result 
        def backtracking(index,sub,total):
            if total == target :
                result.append(sub.copy())
            #If range of index greater than len(candicates) or total > target then return (back to previous case)
            if index>=len(candicates) or total > target:
                return
            #Appending candicates at index into sub 
            #i=0 -> [2]
            sub.append(candicates[index])
            #Inspecting next case
            backtracking(index,sub,total +candicates[index])
            sub.pop()
           # Skip the current candidate and move to the n 
            backtracking(index+1,sub,total)
        backtracking(0,[],0)
        return result
    
    
def test_combinationSum():
    sol = Solution()
    
    # Test case 1
    candidates = [2, 3, 6, 7]
    target = 7
    expected_output = [[2, 2, 3], [7]]
    assert sol.combinationSum(candidates, target) == expected_output, f"Test case 1 failed: {sol.combinationSum(candidates, target)}"
    
    # Test case 2
    candidates = [2, 3, 5]
    target = 8
    expected_output = [[2, 2, 2, 2], [2, 3, 3], [3, 5]]
    assert sol.combinationSum(candidates, target) == expected_output, f"Test case 2 failed: {sol.combinationSum(candidates, target)}"
    
    # Test case 3
    candidates = [2]
    target = 1
    expected_output = []
    assert sol.combinationSum(candidates, target) == expected_output, f"Test case 3 failed: {sol.combinationSum(candidates, target)}"
    
    # Test case 4
    candidates = [1]
    target = 1
    expected_output = [[1]]
    assert sol.combinationSum(candidates, target) == expected_output, f"Test case 4 failed: {sol.combinationSum(candidates, target)}"
    
    # Test case 5
    candidates = [1]
    target = 2
    expected_output = [[1, 1]]
    assert sol.combinationSum(candidates, target) == expected_output, f"Test case 5 failed: {sol.combinationSum(candidates, target)}"
    
    print("All test cases passed!")

# Run tests
test_combinationSum()
